data/transactions/logic/eval.go

```diff
@@ -3154,3 +3206,165 @@ func opAppParamsGet(cx *evalContext) {
+func (cx *evalContext) available_asset(sv stackValue) (basics.AssetIndex, error) {
+	aid, err := sv.uint()
+	if err != nil {
+		return basics.AssetIndex(0), err
+	}
+	// Ensure that aid is in Foreign Assets
+	for _, assetID := range cx.Txn.Txn.ForeignAssets {
+		if assetID == basics.AssetIndex(aid) {
+			return basics.AssetIndex(aid), nil
+		}
+	}
+	return basics.AssetIndex(0), fmt.Errorf("invalid Asset reference %d", aid)
+}
+
+func opTxField(cx *evalContext) {
+	last := len(cx.stack) - 1
+	field := TxnField(uint64(cx.program[cx.pc+1]))
+	sv := cx.stack[last]
+	var i uint64
+	switch field {
+	case Sender:
+		cx.subtxn.Sender, cx.err = cx.available_address(sv)
+	case Receiver:
+		cx.subtxn.Receiver, cx.err = cx.available_address(sv)
+	case Amount:
+		cx.subtxn.Amount.Raw, cx.err = sv.uint()
+	case CloseRemainderTo:
+		cx.subtxn.CloseRemainderTo, cx.err = cx.available_address(sv)
+	case Type:
+		cx.subtxn.Type = protocol.TxType(sv.Bytes)
+	case TypeEnum:
+		i, cx.err = sv.uint()
+		if i < uint64(len(TxnTypeNames)) {
+			cx.subtxn.Type = protocol.TxType(TxnTypeNames[i])
+		}
+	case XferAsset:
+		cx.subtxn.XferAsset, cx.err = cx.available_asset(sv)
+	case AssetReceiver:
+		cx.subtxn.AssetReceiver, cx.err = cx.available_address(sv)
+	case AssetAmount:
+		cx.subtxn.AssetAmount, cx.err = sv.uint()
+	case AssetCloseTo:
+		cx.subtxn.AssetCloseTo, cx.err = cx.available_address(sv)
+	// Keeping things buttoned up for now
+	// Disallow rekey, explicit fee, fv, lv, lease (and everything outside pay, axfr)
+	default:
+		cx.err = fmt.Errorf("invalid txfield %s", field)
+	}
+
+	cx.stack = cx.stack[:last] // pop
+}
+
+func opTxSubmit(cx *evalContext) {
+	if cx.Ledger == nil {
+		cx.err = fmt.Errorf("ledger not available")
+		return
+	}
+
+	// Error out on anything unusual.  Allow pay, axfer.
+	switch cx.subtxn.Type {
+	case protocol.PaymentTx, protocol.AssetTransferTx:
+		// only pay and axfer for now
+	default:
+		cx.err = fmt.Errorf("Invalid action type %s", cx.subtxn.Type)
+		return
+	}
+
+	// The goal is to follow the same invariants used by the
+	// transaction pool. Namely that any transaction that makes it
+	// to Perform (which is equivalent to eval.applyTransaction)
+	// is authorized, and WellFormed.
+	if !authorizedSender(cx, cx.subtxn.Sender) {
+		cx.err = fmt.Errorf("unauthorized")
+		return
+	}
+
+	// Recall that WellFormed does not care about individual
+	// transaction fees because of fee pooling. So we check below.
+	cx.err = cx.subtxn.WellFormed(*cx.Specials, *cx.Proto)
+	if cx.err != nil {
+		return
+	}
+
+	paid := cx.subtxn.Fee.Raw
+	if paid >= cx.Proto.MinTxnFee {
+		// Over paying - accumulate into OverpaidFee
+		overpaid := paid - cx.Proto.MinTxnFee
+		if cx.OverpaidFee == nil {
+			cx.OverpaidFee = new(uint64)
+		}
+		*cx.OverpaidFee = basics.AddSaturate(*cx.OverpaidFee, overpaid)
+	} else {
+		underpaid := cx.Proto.MinTxnFee - paid
+		// Try to pay with OverpaidFee, else fail.
+		if cx.OverpaidFee != nil && *cx.OverpaidFee > underpaid {
```

Comment:
not sure if I'm reading it correctly, but I think that you meant `*cx.OverpaidFee >= underpaid`, right ?

Reply:
Correct.  I'm happy to report that left this here as a reminder to look later , but found the bug with a test before I got around to it.